# Restaurant Recommendation System
# Data Cleaning and Preparation

In [2]:
import pandas as pd
import numpy as np

In [3]:
import matplotlib.pyplot as plt
%matplotlib inline

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
business = pd.read_csv('../data/yelp_data/yelp_business.csv')
user = pd.read_csv('../data/yelp_data/yelp_review.csv')

### Business df

In [125]:
business['business_ref']=pd.factorize(business['business_id'].tolist())[0]

In [126]:
print(business.shape)
business.head()

(174567, 14)


business_id                        name neighborhood  \
0  FYWN1wneV18bWNgQjJ2GNg          "Dental by Design"          NaN   
1  He-G7vWjzVUysIKrfNbPUQ       "Stephen Szabo Salon"          NaN   
2  KQPW8lFf1y5BT2MxiSZ3QA     "Western Motor Vehicle"          NaN   
3  8DShNS-LuFqpEWIp0HxijA          "Sports Authority"          NaN   
4  PfOCPjBrlQAnz__NXj9h_w  "Brick House Tavern + Tap"          NaN   

                            address            city state postal_code  \
0        "4855 E Warner Rd, Ste B9"       Ahwatukee    AZ       85044   
1              "3101 Washington Rd"        McMurray    PA       15317   
2          "6025 N 27th Ave, Ste 1"         Phoenix    AZ       85017   
3  "5000 Arizona Mills Cr, Ste 435"           Tempe    AZ       85282   
4                    "581 Howe Ave"  Cuyahoga Falls    OH       44221   

    latitude   longitude  stars  review_count  is_open  \
0  33.330690 -111.978599    4.0            22        1   
1  40.291685  -80.104900    3.0            11        1   
2  33.524903 -112.115310    1.5            18        1   
3  33.383147 -111.964725    3.0             9        0   
4  41.119535  -81.475690    3.5           116        1   

                                          categories  business_ref  
0  Dentists;General Dentistry;Health & Medical;Or...             0  
1  Hair Stylists;Hair Salons;Men's Hair Salons;Bl...             1  
2  Departments of Motor Vehicles;Public Services ...             2  
3                            Sporting Goods;Shopping             3  
4  American (New);Nightlife;Bars;Sandwiches;Ameri...             4

In [147]:
#We need to filter the business dataframe for Edinburgh and Restaurants for use in the modelling notebooks. 
edin_rest = business[business.city.str.contains('Edin', case=False, na=False)]
edin_rest = edin_rest[edin_rest.categories.str.contains('Rest', case=False, na=False)]
edin_rest.head()

business_id                   name neighborhood  \
68   F31RycVVooeIOp9jsXmg6g    "The Bluebird Cafe"  Cannonmills   
380  inaACfObL1NBNJmBG11iuQ          "Global Deli"  Grassmarket   
397  Di5ApLgoQpcv5Aew82fI_A       "The Rendezvous"     West End   
420  OvbLKXkJCg8ZMHX9L5faIA    "Bread Meats Bread"     West End   
446  T2jfXhvQPk9wLdt1OVV-Kg  "Rose Street Brewery"     New Town   

                             address       city state postal_code   latitude  \
68                    "5 Canonmills"  Edinburgh   MLN     EH3 5HA  55.962444   
380  "13 George IV Bridge, Old Town"  Edinburgh   EDH     EH1 1EE  55.947960   
397          "10 Queensferry Street"  Edinburgh   EDH     EH2 4PG  55.950550   
420                "92 Lothian Road"  Edinburgh   EDH     EH3 9BE  55.946390   
446              "55-57 Rose Street"  Edinburgh   EDH     EH2 2NH  55.952835   

     longitude  stars  review_count  is_open  \
68   -3.197662    4.5            16        1   
380  -3.192143    4.0            13        1   
397  -3.209454    3.5            19        1   
420  -3.205829    4.0            28        1   
446  -3.198560    3.0            24        1   

                                            categories  business_ref  
68   Breakfast & Brunch;Diners;Restaurants;Cafes;Br...            68  
380     Restaurants;Food;Sandwiches;Coffee & Tea;Delis           380  
397                      Restaurants;Cantonese;Chinese           397  
420                                Burgers;Restaurants           420  
446  Pubs;Whiskey Bars;Nightlife;Breakfast & Brunch...           446

In [148]:
edin_rest.shape

(1605, 14)

In [150]:
edin_rest.to_csv('new_restaurants.csv')

### User DataFrame

In [119]:
print(user.shape)
user.head()

(5261668, 9)


review_id                 user_id             business_id  \
0  vkVSCC7xljjrAI4UGfnKEQ  bv2nCi5Qv5vroFiqKGopiw  AEx2SYEUJmTxVVB18LlCwA   
1  n6QzIUObkYshz4dz2QRJTw  bv2nCi5Qv5vroFiqKGopiw  VR6GpWIda3SfvPC-lg9H3w   
2  MV3CcKScW05u5LVfF6ok0g  bv2nCi5Qv5vroFiqKGopiw  CKC0-MOWMqoeWf6s-szl8g   
3  IXvOzsEMYtiJI0CARmj77Q  bv2nCi5Qv5vroFiqKGopiw  ACFtxLv8pGrrxMm6EgjreA   
4  L_9BTb55X0GDtThi6GlZ6w  bv2nCi5Qv5vroFiqKGopiw  s2I_Ni76bjJNK9yG60iD-Q   

   stars        date                                               text  \
0      5  2016-05-28  Super simple place but amazing nonetheless. It...   
1      5  2016-05-28  Small unassuming place that changes their menu...   
2      5  2016-05-28  Lester's is located in a beautiful neighborhoo...   
3      4  2016-05-28  Love coming here. Yes the place always needs t...   
4      4  2016-05-28  Had their chocolate almond croissant and it wa...   

   useful  funny  cool  
0       0      0     0  
1       0      0     0  
2       0      0     0  
3       0      0     0  
4       0      0     0

In [121]:
user['user_ref'] = pd.factorize(user['user_id'].tolist())[0]


In [124]:
user.head(10)

review_id                 user_id             business_id  \
0  vkVSCC7xljjrAI4UGfnKEQ  bv2nCi5Qv5vroFiqKGopiw  AEx2SYEUJmTxVVB18LlCwA   
1  n6QzIUObkYshz4dz2QRJTw  bv2nCi5Qv5vroFiqKGopiw  VR6GpWIda3SfvPC-lg9H3w   
2  MV3CcKScW05u5LVfF6ok0g  bv2nCi5Qv5vroFiqKGopiw  CKC0-MOWMqoeWf6s-szl8g   
3  IXvOzsEMYtiJI0CARmj77Q  bv2nCi5Qv5vroFiqKGopiw  ACFtxLv8pGrrxMm6EgjreA   
4  L_9BTb55X0GDtThi6GlZ6w  bv2nCi5Qv5vroFiqKGopiw  s2I_Ni76bjJNK9yG60iD-Q   
5  HRPm3vEZ_F-33TYVT7Pebw  _4iMDXbXZ1p1ONG297YEAQ  8QWPlVQ6D-OExqXoaD2Z1g   
6  ymAUG8DZfQcFTBSOiaNN4w  u0LXt3Uea_GidxRW1xcsfg  9_CGhHMz8698M9-PkVf0CQ   
7  8UIishPUD92hXtScSga_gw  u0LXt3Uea_GidxRW1xcsfg  gkCorLgPyQLsptTHalL61g   
8  w41ZS9shepfO3uEyhXEWuQ  u0LXt3Uea_GidxRW1xcsfg  5r6-G9C4YLbC7Ziz57l3rQ   
9  WF_QTN3p-thD74hqpp2j-Q  u0LXt3Uea_GidxRW1xcsfg  fDF_o2JPU8BR1Gya--jRIA   

   stars        date                                               text  \
0      5  2016-05-28  Super simple place but amazing nonetheless. It...   
1      5  2016-05-28  Small unassuming place that changes their menu...   
2      5  2016-05-28  Lester's is located in a beautiful neighborhoo...   
3      4  2016-05-28  Love coming here. Yes the place always needs t...   
4      4  2016-05-28  Had their chocolate almond croissant and it wa...   
5      5  2014-09-24  Cycle Pub Las Vegas was a blast! Got a groupon...   
6      4  2012-05-11  Who would have guess that you would be able to...   
7      4  2015-10-27  Always drove past this coffee house and wonder...   
8      3  2013-02-09  Not bad!! Love that there is a gluten-free, ve...   
9      5  2016-04-06  Love this place!\n\nPeggy is great with dogs a...   

   useful  funny  cool  user_ref  
0       0      0     0         0  
1       0      0     0         0  
2       0      0     0         0  
3       0      0     0         0  
4       0      0     0         0  
5       1      0     0         1  
6       0      0     2         2  
7       1      0     0         2  
8       1      0     0         2  
9       3      0     0         2

For our initialcollaborative based models we'll require only a few columns, and some initially just for clarity. We want a dataset that is Edinburgh centric, based on restaurants only and with the user_id, business_id and stars as the features. Note that we caanot just use these columns from the user dataframe as there is no way of filtering by location and business type without joining the business dataset. So this is what we'll do first. 

### Select required features from each dataset

In [129]:
#business_id, city, name and categories from the business dataset
bus_cut = business[['business_ref', 'business_id', 'name', 'city', 'categories']]
bus_cut.head()

business_ref             business_id                        name  \
0             0  FYWN1wneV18bWNgQjJ2GNg          "Dental by Design"   
1             1  He-G7vWjzVUysIKrfNbPUQ       "Stephen Szabo Salon"   
2             2  KQPW8lFf1y5BT2MxiSZ3QA     "Western Motor Vehicle"   
3             3  8DShNS-LuFqpEWIp0HxijA          "Sports Authority"   
4             4  PfOCPjBrlQAnz__NXj9h_w  "Brick House Tavern + Tap"   

             city                                         categories  
0       Ahwatukee  Dentists;General Dentistry;Health & Medical;Or...  
1        McMurray  Hair Stylists;Hair Salons;Men's Hair Salons;Bl...  
2         Phoenix  Departments of Motor Vehicles;Public Services ...  
3           Tempe                            Sporting Goods;Shopping  
4  Cuyahoga Falls  American (New);Nightlife;Bars;Sandwiches;Ameri...

In [130]:
users_cut = user[['business_id', 'review_id', 'user_id', 'stars', 'user_ref']]
users_cut.head()

business_id               review_id                 user_id  \
0  AEx2SYEUJmTxVVB18LlCwA  vkVSCC7xljjrAI4UGfnKEQ  bv2nCi5Qv5vroFiqKGopiw   
1  VR6GpWIda3SfvPC-lg9H3w  n6QzIUObkYshz4dz2QRJTw  bv2nCi5Qv5vroFiqKGopiw   
2  CKC0-MOWMqoeWf6s-szl8g  MV3CcKScW05u5LVfF6ok0g  bv2nCi5Qv5vroFiqKGopiw   
3  ACFtxLv8pGrrxMm6EgjreA  IXvOzsEMYtiJI0CARmj77Q  bv2nCi5Qv5vroFiqKGopiw   
4  s2I_Ni76bjJNK9yG60iD-Q  L_9BTb55X0GDtThi6GlZ6w  bv2nCi5Qv5vroFiqKGopiw   

   stars  user_ref  
0      5         0  
1      5         0  
2      5         0  
3      4         0  
4      4         0

In [131]:
bus_revs = bus_cut.merge(users_cut, how='left', left_on='business_id', right_on='business_id')
bus_revs.head()

business_ref             business_id                name       city  \
0             0  FYWN1wneV18bWNgQjJ2GNg  "Dental by Design"  Ahwatukee   
1             0  FYWN1wneV18bWNgQjJ2GNg  "Dental by Design"  Ahwatukee   
2             0  FYWN1wneV18bWNgQjJ2GNg  "Dental by Design"  Ahwatukee   
3             0  FYWN1wneV18bWNgQjJ2GNg  "Dental by Design"  Ahwatukee   
4             0  FYWN1wneV18bWNgQjJ2GNg  "Dental by Design"  Ahwatukee   

                                          categories               review_id  \
0  Dentists;General Dentistry;Health & Medical;Or...  3_-fne2m2GvKESZRlULAiw   
1  Dentists;General Dentistry;Health & Medical;Or...  5gJk64TDq5lqSuoNiaqjVQ   
2  Dentists;General Dentistry;Health & Medical;Or...  5570rwX_o30Z3NmKhou4CQ   
3  Dentists;General Dentistry;Health & Medical;Or...  iRm438GDAajUiW2GoIAvtA   
4  Dentists;General Dentistry;Health & Medical;Or...  9lhJwhRwTsqqPyyoSjMuqw   

                  user_id  stars  user_ref  
0  jTvKkNTat9QftSxL_FMcQw      5     33075  
1  IG6XMgQyWF2_RW1nKEioxA      5     80363  
2  2Rh41EBjWI137N2m5VHJJA      5    140780  
3  YcmNpPM0ag94g4T0zAtdcg      4    162116  
4  XGL7VDkeUyM5nKQspJBTNw      5    194603

In [132]:
bus_revs.shape

(5261668, 9)

### Filter for our required location and business type

Now we know that we have the individual user ratings and the businesses merged we can filter for Edinburgh and Restaurants / cafes. 

In [133]:
edin_revs = bus_revs[bus_revs.city.str.contains('Edin', case=False, na=False)]

In [134]:
print(edin_revs.shape)
edin_revs.head()

(52492, 9)


business_ref             business_id                 name       city  \
1900            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1901            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1902            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1903            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1904            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   

                                             categories  \
1900  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1901  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1902  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1903  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1904  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   

                   review_id                 user_id  stars  user_ref  
1900  b31UZTy2TvnFtkfygJG40Q  bcxcQhp0sKYd9eUnEVUzPA      5       246  
1901  jYxWLyWrWy8dJFQs9DEuEg  RFxjYeLW_aYLdVW3PBwFNg      4    146572  
1902  DMXFC1hZj_e4Mb7iOlVsGg  Lo6hBSAP4grz6Cnr3xiCKw      5    196666  
1903  GGWxoYbx_h2x7a46m0MYRA  BhYROfCjIJsKUk22_IVHig      3    211139  
1904  PslbThtGZ_yOWZxAFc3GVg  J_qpI2jCkwv7vPNz_9JeqA      4    260271

In [135]:
edin_rest_revs = edin_revs[edin_revs.categories.str.contains('restaurants', case=False, na=False)]

In [136]:
print(edin_rest_revs.shape)
edin_rest_revs.head()

(28407, 9)


business_ref             business_id                 name       city  \
1900            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1901            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1902            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1903            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1904            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   

                                             categories  \
1900  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1901  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1902  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1903  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1904  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   

                   review_id                 user_id  stars  user_ref  
1900  b31UZTy2TvnFtkfygJG40Q  bcxcQhp0sKYd9eUnEVUzPA      5       246  
1901  jYxWLyWrWy8dJFQs9DEuEg  RFxjYeLW_aYLdVW3PBwFNg      4    146572  
1902  DMXFC1hZj_e4Mb7iOlVsGg  Lo6hBSAP4grz6Cnr3xiCKw      5    196666  
1903  GGWxoYbx_h2x7a46m0MYRA  BhYROfCjIJsKUk22_IVHig      3    211139  
1904  PslbThtGZ_yOWZxAFc3GVg  J_qpI2jCkwv7vPNz_9JeqA      4    260271

In [137]:
#edin_rest_revs.to_csv('edinburgh_reviews.csv')

In [138]:
edin_rest_revs.business_id.nunique()

1605

In [139]:
edin_rest_revs.user_id.nunique()

9162

### User reviews analysis

Let's see how the user reveiws are distributed among the user base... we may need to make a decision about removing users with a low review counts. 

In [140]:
edin_rest_revs['user_id'].value_counts().to_frame().describe()

user_id
count  9162.000000
mean      3.100524
std      10.324198
min       1.000000
25%       1.000000
50%       1.000000
75%       2.000000
max     272.000000

In [141]:
user_counts = edin_rest_revs['user_id'].value_counts()
user_counts

utRlrauZEOF_2kASsIoJ7A    272
txzOZrHK9EVY0kYjSuXkKg    243
LURC3E0DoXYgN9aYTF3XOg    224
Cy1PV2TdYeDFA16mUOlr3Q    205
yfXqZkU5iXE07GSHzdsQBA    204
                         ... 
wSKp3SBOA3pytkwYXtB6rQ      1
VKgu2AoBHLbOY2zSfphvpA      1
nfcBvVbzRqVU_qoq423gtg      1
fuwedVoOZASUk4asvnxElw      1
7bebusNNG4-oiuqMiPFwPQ      1
Name: user_id, Length: 9162, dtype: int64

In [142]:
user_list = user_counts[user_counts > 10].index.tolist()
len(user_list)

348

so only 348 of our 9162 users have posted more than 10 reviews, when it comes to vectorizing then the matrix is going to be very sparse indeed. This may impact the predictive power of the final recommender. 

In [143]:
edin_rest_revs = edin_rest_revs[edin_rest_revs['user_id'].isin(user_list)]
edin_rest_revs.head()

business_ref             business_id                 name       city  \
1900            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1901            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1903            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1904            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   
1907            68  F31RycVVooeIOp9jsXmg6g  "The Bluebird Cafe"  Edinburgh   

                                             categories  \
1900  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1901  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1903  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1904  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   
1907  Breakfast & Brunch;Diners;Restaurants;Cafes;Br...   

                   review_id                 user_id  stars  user_ref  
1900  b31UZTy2TvnFtkfygJG40Q  bcxcQhp0sKYd9eUnEVUzPA      5       246  
1901  jYxWLyWrWy8dJFQs9DEuEg  RFxjYeLW_aYLdVW3PBwFNg      4    146572  
1903  GGWxoYbx_h2x7a46m0MYRA  BhYROfCjIJsKUk22_IVHig      3    211139  
1904  PslbThtGZ_yOWZxAFc3GVg  J_qpI2jCkwv7vPNz_9JeqA      4    260271  
1907  oRYhx_qYK5slteB5nyEAiQ  NMelfYHO9mncdmZLIABLgQ      5    491834

In [144]:
edin_rest_revs.shape

(12552, 9)

In [145]:
edin_rest_revs.business_id.nunique()

1527

To enable a base engine to be built we will progress with this dataframe as our data set for collaborative filtering. From the initially imported Yelp datasets we have filered by categories = 'restaurant', location = Edinburgh, and minimum-user-reviews = 10. This leaves us with 12552 reviews across 348 users and 1527 business. We note that removing users with low review counts has also removed some businesses from the set - logically this suggests that the removed restuarants could also have low review counts which could be detrimental to our final product as we want to be able to get access to these lesser known establishments.  

Let's export two data sets for collaborative modelling purposes. One that sets a lower limit for user reviews at 10 and one which sets no lower limit for comparison. 

In [146]:
edin_rest_revs.to_csv('new_collaborative_limit.csv')